scripting/conn-id-ctx.rst

```diff
@@ -0,0 +1,103 @@
+
+.. _script-conn-id-ctx:
+
+==============================
+Use of :zeek:see:`conn_id_ctx`
+==============================
+
+.. versionadded:: 8.0
+
+.. note::
+
+   We’re still iterating on patterns for working with the new pluggable
+   connection keys and :zeek:see:`conn_id_ctx` instances.
+   If you have feedback or run into limitations for your use-cases, please reach out!
+
+In certain deployments, Zeek will receive network traffic from different
+network segments that may have overlapping IP ranges.
+For example, host 10.0.0.37 in segment A and host 10.0.0.37 in segment B
+share the same IP address, but represent different systems.
+The terminology used within Zeek is that IP addresses or connections
+are observed in different "contexts".
+
+
+Commonly, ethernet layers may contain information that makes it possible to discriminate
+between different network segments. Concretely, VLAN tagging is one such approach.
+Segments can also be distinguished by Virtual Network Identifiers (VNIs)
+in case of UDP-based tunnels like VXLAN or Geneve.
+
+
+Since Zeek 8.0, this information can be extracted by
+:ref:`plugin-provided connection key implementations <connkey-plugin>`
+and included into Zeek's core connection tracking. Further, plugins
+:zeek:keyword:`redefine <redef>` :zeek:see:`conn_id_ctx` with additional
+fields to expose the extracted information used for connection tracking to
+the Zeek scripting layer.
+For example, loading the doc:`/scripts/policy/frameworks/conn_key/vlan_fivetuple.zeek`
+adds :zeek:field:`vlan` and :zeek:field:`inner_vlan` fields to :zeeK:see:`conn_id_ctx`.
+
+Script writers can use the :zeek:field:`conn_id$ctx <conn_id$ctx>` field to
+discriminate the same :zeek:type:`addr` values observed in different contexts.
+
+For example, to count the number of connections per originator address in
+a context-aware manner, add the :zeek:see:`conn_id_ctx` instance as part
+of the index into a table:
+
+.. code-block:: zeek
+
+	global connection_counts: table[conn_id_ctx, addr] of count &default=0;
+
+	event new_connection(c: connection) {
+		++connection_counts[c$id$ctx, c$id$orig_h];
+	}
+
+
+If :zeek:field:`ctx` is populated with fields for VLAN tags, the table indexing
+will take them into account and create individual entries per ``(ctx, addr)``
+pair.
```

Comment:
```suggestion
If, for example, :zeek:field:`ctx` is populated with fields for VLAN tags, that table will create individual entries per ``(VLAN, addr)`` pair. This will also work correctly if no context has been defined: `c$id$ctx` will then just be an empty records with no fields. (This remains the default for most Zeek installations.)
```